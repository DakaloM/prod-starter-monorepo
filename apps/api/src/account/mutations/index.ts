export { createUser } from './createUser';
export { recoverPassword } from './recoverPassword';
export { forgotPassword } from './forgotPassword';
export * from './updateUser';
// export * from './userRegister'

