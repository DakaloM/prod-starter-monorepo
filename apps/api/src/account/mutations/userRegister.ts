// import { UserRegisterInput } from "@imax/client";
// import { BadRequestError, ConflictError } from "@imax/serverkit";
// import { faker } from "@imax/testkit";
// import * as argon from "argon2";
// import * as saIdParser from "south-african-id-parser";
// import { Context } from "~/context";
// import { getMatchValues } from "~/domain/match";
// import { Applicant } from "../applicant";
// import { User, UserRole } from "../user";
// import { Contact } from "~/contact";
// import { createContact } from "~/contact/mutation";

// export async function userRegister(input: UserRegisterInput, ctx: Context) {
//   const { confirmPassword, mobileNumber, businessNumber, privateNumber, idNumber, ...attrs } =
//     input;
//   const { email, password, ...rest } = attrs;

//   const user = await User.query(ctx.db)
//     .where({ email })
//     .orWhere({ idNumber })
//     .first();

//   if (user) {
//     throw new ConflictError({
//       message: "User already exist",
//       path: getMatchValues(user, { email }),
//     });
//   }

//   if (password !== confirmPassword) {
//     throw new BadRequestError({
//       message: "Passwords do not match",
//     });
//   }

//   const parseInfo = saIdParser.parse(idNumber);
//   if (!parseInfo.isValid) {
//     throw new BadRequestError({ message: "Invalid RSA ID Number" });
//   }

//   const sortName = User.getSortName(attrs.name, attrs.surname);

//   return await User.transaction(ctx.db, async (trx) => {
//     const newUser = await User.query(trx).insert({
//       ...attrs,
//       password: await argon.hash(password),
//       sortName,
//       role: UserRole.Applicant,
//       idNumber
//     });

//     await Applicant.query(trx).insert({
//       userId: newUser.id,
//       applicantNumber: faker.string.numeric(6),
//       name: newUser.name,
//       surname: newUser.surname,
//     });

//     await Contact.query(trx).insert({
//       userId: newUser.id,
//       email,
//       mobileNumber,
//       businessNumber,
//       privateNumber,
//     });

//     return newUser;
//   });
// }
