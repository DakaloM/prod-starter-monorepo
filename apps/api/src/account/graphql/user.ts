import { builder } from '~/graphql/builder';
import { isActiveUser, isDashboardOperator } from '~/account/shield';
import { Gender, UserRole, UserStatus, Race, Title, User } from '~/account/user';
import { nullable } from 'zod';
import { getUserById } from '../queries';
import { allow, or } from 'graphql-shield';


export const UserRoleSchema = builder.enumType(UserRole, {
  name: 'UserRole',
  description: 'User role',
});

export const getEnumValues = <T extends Record<string, any>>(enumType: T) => {
  return Object.keys(enumType).reduce((acc, key) => {
    return {
      ...acc,
      [key]: { value: enumType[key] },
    };
  }, {});
};


export const GenderSchema = builder.enumType(Gender, {
  name: 'Gender',
  description: 'Gender',
});

export const RaceSchema = builder.enumType(Race, {
  name: 'Race',
  description: 'Race',
});

export const TitleSchema = builder.enumType(Title, {
  name: 'Title',
  description: 'Title',
});

export const UserStatusSchema = builder.enumType(UserStatus, {
  name: 'UserStatus',
  description: 'User status',
});

export const UserSchema = builder.objectType(User, {
  name: 'User',
  shield: allow,
  description: 'A user is an human user or bot that can interact with the system.',

  fields: (t) => ({
    id: t.exposeID('id'),
    email: t.field({
      type: 'EmailAddress',
      resolve: (root) => root.email,
    }),
    role: t.expose('role', { type: UserRoleSchema, description: 'The role of the user.' }),
    name: t.exposeString('name'),
    surname: t.exposeString('surname'),
    middleName: t.exposeString('middleName',{nullable: true}),
    title: t.exposeString('title' ,{nullable: true}),
    gender: t.exposeString('gender' ,{nullable: true}),
    idNumber: t.exposeString('idNumber'),
    race: t.exposeString('race' ,{nullable: true}),
    status: t.exposeString('status'),
    birthDate: t.expose('birthDate', {type: 'Date', nullable: true}),
    sequence: t.exposeInt('sequence'),
    applicantId: t.exposeString('applicantId', {nullable: true})
   
  }),
  
});

builder.queryField('user', (t) =>
  t.field({
    shield: or(isActiveUser, isDashboardOperator),
    description: 'A user',
    args: {
      id: t.arg.id({ required: true }),
    },
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      return getUserById(args.id.toString(), ctx);
    },
  }),
);

builder.queryField('me', (t) =>
  t.field({
    shield: isActiveUser,
    description: 'A user',
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      const id = ctx.auth.user?.id.toString() || '';

      return getUserById(id, ctx);
    },
  }),
);




