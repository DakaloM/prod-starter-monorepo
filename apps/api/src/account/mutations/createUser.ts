import { CreateUserInput } from '@imax/client';
import { BadRequestError, ConflictError, NotFoundError } from '@imax/serverkit';
import { faker } from '@imax/testkit';

import { Knex } from 'knex';
import * as saIdParser from 'south-african-id-parser';
import { Context } from '~/context';
import { getMatchValues } from '~/domain/match';
import { Gender, UserRole, Race, Title, User, UserStatus } from '../user';

export async function createUser(input: CreateUserInput, ctx: Context, trx?: Knex) {
  const db = trx || ctx.db;
  const { email, idNumber, middleName, name, surname, birthDate } = input;

  const role = input.role as UserRole;
  const title = input.title as Title;
  const race = input.race as Race;
  const gender = input.gender as Gender;

  const existingUser = await User.query(db).where({ email }).orWhere({ idNumber }).first();

  if (existingUser) {
    throw new ConflictError({
      message: 'User already exists',
      path: getMatchValues(existingUser, { idNumber, email }),
    });
  }

  const parseInfo = saIdParser.parse(idNumber);
  if (!parseInfo.isValid) {
    throw new BadRequestError({ message: 'Invalid RSA ID Number' });
  }

  const sortName = User.getSortName(input.name, input.surname);

  return User.query(db).insert({
    email,
    idNumber,
    name,
    surname,
    middleName: middleName ? middleName : undefined,
    gender,
    race,
    title,
    birthDate,
    sortName,
    role,
    status: UserStatus.Active,
  });
}


