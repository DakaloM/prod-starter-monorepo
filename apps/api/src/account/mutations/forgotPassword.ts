import { ExpiresInSeconds } from '@imax/cache';
import { ForbiddenError } from '@imax/serverkit';

import { User } from '~/account/user';
import { Context } from '~/context';

type Args = {
  email: string;
};

export async function forgotPassword({ email }: Args, ctx: Context) {
  const user = await User.query(ctx.db).findOne({
    email,
  });

  if (!user) {
    throw new ForbiddenError({});
  }
  const [, k] = Math.random().toString().split('.');
  const code = k.slice(0, 6);

  await ctx.cache.set(email, code, { amount: 5, value: ExpiresInSeconds.OneMinute });

  await ctx.mailer.sendMail({
    to: user.email,
    subject: 'Forgot password',
    data: {
      title: 'Forgot password',
      subTitle: '',
      paragraphs: [`Your code is <strong>${code}</strong>`],
    },
  });

  return '';
 
}
