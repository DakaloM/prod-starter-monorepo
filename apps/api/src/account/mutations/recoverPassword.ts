import { ForbiddenError } from '@imax/serverkit';

import * as argon from 'argon2';
import { User } from '~/account/user';
import { createAuthTokens } from '~/auth/mutation/createAuthTokens';
import { Context } from '~/context';

type Args = {
  email: string;
  code: string;
  password: string;
};

export async function recoverPassword(args: Args, ctx: Context) {
  const { code, password, email } = args;

  const user = await User.query(ctx.db).findOne({
    email,
  });

  if (!user) {
    throw new ForbiddenError({});
  }

  const savedCode = await ctx.cache.get(email);

  const isValid = code === savedCode;

  if (!isValid) {
    throw new ForbiddenError({});
  }

  await ctx.cache.remove(user.email);

  await User.query(ctx.db)
    .findById(user.id)
    .patch({
      password: await argon.hash(password),
    });

  return createAuthTokens(user, ctx);
}
