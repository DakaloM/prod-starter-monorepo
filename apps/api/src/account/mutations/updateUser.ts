import { Gender, Race, Title, UserStatus } from '~/account';
import { BadRequestError, NotFoundError } from '@imax/serverkit';

import * as saIdParser from 'south-african-id-parser';
import { Context } from '~/context';

import { User } from '../user';
import { UpdateUserInput } from '@imax/client';

export async function updateUser(input: UpdateUserInput, ctx: Context) {
  const { id, ...attrs } = input;
  const idNumber = attrs.idNumber ? attrs.idNumber : undefined;
  const name = attrs.name ? attrs.name : undefined;
  const surname = attrs.surname ? attrs.surname : undefined;

  const db = ctx.db;
  const user = await User.query(db).findById(id);

  if (!user) {
    throw new NotFoundError({
      message: 'User not found',
    });
  }

  if (idNumber) {
    const parseInfo = saIdParser.parse(idNumber);
    if (!parseInfo.isValid) {
      throw new BadRequestError({ message: 'Invalid RSA ID Number' });
    }
  }

  const sortName = User.getSortName(name ? name : user.name, surname ? surname : user.surname);
  const change = JSON.parse(JSON.stringify(attrs));
  const patchInput = {
    ...change,
    sortName: sortName,
  };

  const updatedUser = await user.$query(db).patchAndFetch(patchInput);

  return updatedUser;
}

