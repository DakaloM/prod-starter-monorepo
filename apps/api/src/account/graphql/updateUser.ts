
import { builder } from "~/graphql/builder";
import { GenderSchema, RaceSchema, TitleSchema, UserRoleSchema, UserSchema, UserStatusSchema } from "./user";
import { isActiveUser } from "../shield";
import { updateUser } from "../mutations";


const UpdateUserInput = builder.inputType('UpdateUserInput', {
  fields: (t) => ({
    id: t.string({required: true}),
    name: t.string({required: false}),
    middleName: t.string({required: false}),
    surname: t.string({required: false}),
    email: t.string({required: false}),
    title: t.field({type: TitleSchema, required: false}),
    race: t.field({type: RaceSchema, required: false}),
    gender: t.field({type: GenderSchema, required: false}),
    birthDate: t.field({type: 'Date', required: false}),
    idNumber: t.string({required: false}),
    role: t.field({type: UserRoleSchema, required: false}),
    status: t.field({type: UserStatusSchema, required: false})
  })
});

builder.mutationField('updateUser', (t) => (
  t.field({
    shield: isActiveUser,
    description: 'Update user',
    args: {
      input: t.arg({type: UpdateUserInput, required: true}),
    },
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      return updateUser(args.input, ctx);
    }
  })
))