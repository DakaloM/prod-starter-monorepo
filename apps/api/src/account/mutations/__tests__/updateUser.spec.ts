
import { faker } from "@imax/testkit";
import { User, UserRole, UserStatus } from "~/account/user";
import { Context } from "~/context";
import { createContext, destroyContext } from "~/test/context";
import { Environment, createEnv, destroyEnv } from "~/test/environment"
import { Factory, createFactory } from "~/test/factory";
import { updateUser } from "~/account/mutations/updateUser";
import { omit } from "lodash";

describe('user/mutation/updateUser', () => {
  let env: Environment;
  let ctx: Context;
  let factory: Factory;
  let user : User;

  beforeAll(async () => {
    env = await createEnv()
  });

  afterAll(async () => {
    await destroyEnv(env);
  });


  beforeEach(async () => {
    ctx = await createContext(env);
    factory = createFactory(ctx);
    await factory.seed('types');
    ctx.auth.update(await factory.insert('user'));
    user = await factory.insert('user', {
      status: UserStatus.Active,
      role: UserRole.Admin
    })
  })

  afterEach(async () => {
    await destroyContext(ctx);
  });

  it('should update a user', async () => {
    const update = {
      id: user.id,
      name: faker.person.firstName(),
      surname: faker.person.lastName(),
      status: UserStatus.Inactive,
      role: UserRole.Applicant
    };

    

    const response = await updateUser(update, ctx);

    const updatedUser = await User.query(ctx.db).findById(user.id).throwIfNotFound();
    expect(updatedUser).toBeDefined();
    expect(updatedUser?.name).toEqual(update.name);
    expect(updatedUser?.status).toEqual(update.status);
    expect(updatedUser?.surname).toEqual(update.surname);
    expect(updatedUser?.role).toEqual(update.role);

    expect(response).toMatchObject(omit(updatedUser, ['search']));

  })

  it('should fail if user does not exist', async () => {
    const update = {
      id: faker.string.uuid(),
      name: faker.person.firstName(),
      surname: faker.person.lastName(),
      status: UserStatus.Inactive,
      role: UserRole.Applicant
    };

    await expect(
      updateUser(update, ctx)
    ).rejects.toMatchSnapshot();
  })


})