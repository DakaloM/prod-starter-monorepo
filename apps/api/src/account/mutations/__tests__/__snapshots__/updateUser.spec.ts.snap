// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user/mutation/updateUser > should fail if user does not exist 1`] = `[Error: User not found]`;
